### PR TITLE
test: Gsheet Automation

### PR DESCRIPTION
**Pull request by sarojsarab**
## Description
- Automated below queries for gsheet

1. Fetch Details
2. Insert one
3. Insert many
4. Update one
5. Update many
6. Fetch many
7. Delete One

- Added workflow to schedule run for gsheet automated tests on the hosted instance
- Added a new cypress config file which will be used to run the gsheet tests
- Added tests for folowing permission/scope options  `All access`, `selected access`, `Read/write | All google sheets` and `Read | All google sheets`
- Added negative scenarios
- Added tests to verify widget binding for both suggested widget and drag n drop widget for both selected and all access permission/scope
#### Type of change
- Cypress
#### How Has This Been Tested?
- [x] Cypress
>
>
#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Review by Aishwarya-U-R — COMMENTED** [review]
Can you update PR description to include the scenarios which are tested in this PR?
And where is sheet exactly authenticated?

**Comment by sarojsarab** [comment]
> Can you update PR description to include the scenarios which are tested in this PR?
> And where is sheet exactly authenticated?

Thanks for the feedback, I have updated the description with what has been covered.

To answer your 2nd question, we are not authenticating gsheet from code as it is not possible, we have hosted appsmith at https://regression.test.appsmith.com and have an existing workspace and authenticated gsheet datasource to run the tests. We will not be running gsheet tests as part of our regular CI runs, I have added a new workflow to run the gsheet tests on a schedule basis.

File: app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  before("Setup app and spreadsheet", function () {
+    //Add a new app and an add new spreadsheet query
+    //Setting up the spreadsheet name
+    const uuid = Cypress._.random(0, 10000);
+    spreadSheetName = spreadSheetName + "_" + uuid;
+    appName = appName + "-" + uuid;
+
+    //Adding query to insert a new spreadsheet
+    homePage.NavigateToHome();
+    homePage.CreateAppInWorkspace(workspaceName);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts:25]: If a workspace in that machine is already have workspace with sheet authorised, why are we creating a new workspace here?

File: app/client/cypress/support/Pages/DataSources.ts [hunk 2]
```diff
@@ -1534,4 +1539,40 @@ export class DataSources {
       });
     });
   }
+
+  public enterSortByValues(sortBy: string, option: string, index = 0) {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:1543]: TS helper methods names are CamelCase - `EnterSortByValues`

File: app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts (hunk 1 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts:25]: We are not creating workspace here, we are creating a new app here in the existing workspace.

File: app/client/cypress/support/Pages/DataSources.ts (hunk 2 above)
Comment (sarojsarab) [reply app/client/cypress/support/Pages/DataSources.ts]: Fixed. Thanks

File: app/client/cypress/fixtures/test-data-gsheet.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
+export const GSHEET_DATA = [
```
Comment (Aishwarya-U-R) [inline app/client/cypress/fixtures/test-data-gsheet.ts:1]: What is use of this file?

File: .github/workflows/ci-test-gsheet.yml [hunk 4]
```diff
@@ -0,0 +1,443 @@
+name: Appsmith CI Gsheet Test Workflow
```
[the remark this answers is not in the document]
Comment (sarojsarab) [reply .github/workflows/ci-test-gsheet.yml]: Updated, thanks!

File: app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    });
+    dataSources.RunQueryNVerifyResponseViews(GSHEET_DATA.length);
+    dataSources.AssertQueryTableResponse(
+      1,
+      "Learning Curve Chuggington Interactive Chatsworth",
+    );
+
+    // Sort by descending and verify
+    dataSources.clearSortByOption(); //clearing previous sort option
+    dataSources.enterSortByValues("price", "Descending");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts:145]: References should also be corrected.

File: app/client/cypress/fixtures/test-data-gsheet.ts (hunk 3 above)
Comment (sarojsarab) [reply app/client/cypress/fixtures/test-data-gsheet.ts:1]: We use this data for insert/update to the spreadsheet

File: app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts (hunk 5 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/GsheetFunctional_Spec.ts]: Corrected the references, thanks!

File: app/client/cypress/e2e/GSheet/GSheetWidgetBinding_spec.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
+/// <reference types="Cypress" />
```
Comment (Aishwarya-U-R) [inline app/client/cypress/e2e/GSheet/GSheetWidgetBinding_spec.ts:1]: Is this line needed?

**Review by Aishwarya-U-R — DISMISSED** [review]
Approving current changes.
Can we merge this PR & start running @sarojsarab ? Any new tests from Prapula can come in later as new PR.

File: app/client/cypress/e2e/GSheet/GSheetWidgetBinding_spec.ts (hunk 6 above)
Comment (sarojsarab) [reply app/client/cypress/e2e/GSheet/GSheetWidgetBinding_spec.ts:1]: I think it's better to be there as is used for type checking and intellisense in Cypress tests for typescript, also it helps the TypeScript compiler to understand the Cypress-specific syntax and provides better code suggestions and error checking.

**Comment by sarojsarab** [comment]
> Approving current changes.
> Can we merge this PR & start running @sarojsarab ? Any new tests from Prapula can come in later as new PR.
Thanks @Aishwarya-U-R 
And yes, let's merge this. I will raise new PR for upcoming changes.

File: app/client/cypress/support/Pages/DataSources.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import { ObjectsRegistry } from "../Objects/Registry";
 import { WIDGET } from "../../locators/WidgetLocators";
 import { EntityItems } from "./AssertHelper";
+import formControl from "../../locators/FormControl.json";
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:4]: Bring the needed locator from formControl to this file & remove this import..

Comment (sarojsarab) [reply app/client/cypress/support/Pages/DataSources.ts]: Done!

File: app/client/cypress/support/Pages/DataSources.ts [hunk 8]
```diff
@@ -231,10 +233,15 @@ export class DataSources {
...
   private _reconnectModalDSToopTipIcon = ".t--ds-list .ads-v2-icon";
+  _multiSelectDropdown = (ddName: string) =>
+    "//p[contains(text(),'" +
+    ddName +
+    "')]/ancestor::div[@class='form-config-top']/following-sibling::div//div[contains(@class, 'rc-select-multiple')]";
   private _datasourceTableSchemaInQueryEditor =
     ".datasourceStructure-query-editor";
   private _datasourceColumnSchemaInQueryEditor = ".t--datasource-column";
   private _datasourceStructureSearchInput = ".datasourceStructure-search input";
+  _JSModeSortingControl = ".t--actionConfiguration\\.formData\\.sortBy\\.data";
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/DataSources.ts:244]: Variable name starts small case` _jsModeSortingControl`

Comment (sarojsarab) [reply app/client/cypress/support/Pages/DataSources.ts]: Done!

**Review by Aishwarya-U-R — APPROVED** [review]

**Pull request merged**